Add doc about how to get self subject attributes

nabokihms opened the pull request:
Documentation for:
<!-- ℹ️

 Hello!

 Remember to ADD A DESCRIPTION and delete this note before submitting
 your pull request. The description should explain what will change,
 and why.

 PLEASE title the FIRST commit appropriately, so that if you squash all
 your commits into one, the combined commit message makes sense.
 For overall help on editing and submitting pull requests, visit:
  https://kubernetes.io/docs/contribute/suggest-improvements/

 Use the default base branch, “main”, if you're documenting existing
 features in the English localization.

 If you're working on a different localization (not English), see
 https://kubernetes.io/docs/contribute/new-content/overview/#choose-which-git-branch-to-use
 for advice.

 If you're documenting a feature that will be part of a future release, see
 https://kubernetes.io/docs/contribute/new-content/new-features/ for advice.

-->

  - [x] KEP (`k/enhancements`) update PR(s): https://github.com/kubernetes/enhancements/pull/3326
  - [x] Code (`k/k`) update PR(s): https://github.com/kubernetes/kubernetes/pull/111333
  - [ ] Docs (`k/website`) update PR(s): https://github.com/kubernetes/website/pull/35385

divya-mohan0209 commented:
@nabokihms Hello there! I'm assuming this is for the 1.25 release. If yes, please could you resubmit the PR against the dev-1.25 branch instead of the main branch?

cathchu commented:
Hey there 👋 @nabokihms , 1.25 Release Docs Shadow here!

Thanks for filing this placeholder. As a reminder, final docs PRs are due August 9th. Take a look at [Documenting a release](https://kubernetes.io/docs/contribute/new-content/new-features/#for-developers-or-other-sig-members) for additional information for the docs requirement for the release.

Thanks again!

Sea-n commented:
/remove-area blog
/remove-language ja ko pt zh

reylejano commented:
/assign @cathchu 
/cc @kcmartin 
/hold
hold on confirming the status of the related enhancement https://github.com/kubernetes/enhancements/issues/3325

reylejano commented:
/milestone clear
@nabokihms , since this enhancement is no longer tracked for 1.25, you can target this PR for `main` and change the target when the `dev-1.26` branch opens when 1.25 is released

Sea-n commented on content/en/docs/reference/access-authn-authz/authentication.md:
```diff
@@ -1239,6 +1239,49 @@ The following `ExecCredential` manifest describes a cluster information sample.
 
+## Get self subject attributes
+
+{{< feature-state for_k8s_version="v1.25" state="alpha" >}}
```
```suggestion
{{< feature-state for_k8s_version="v1.26" state="alpha" >}}
```

Sea-n reviewed (commented):
It's [targeting v1.26](https://github.com/kubernetes/enhancements/issues/3325#issuecomment-1234293022) now.

And please change the target branch to `dev-1.26`.

sftim commented:
/sig auth

sftim commented on content/en/docs/reference/access-authn-authz/authentication.md:
```diff
@@ -1239,6 +1239,49 @@ The following `ExecCredential` manifest describes a cluster information sample.
 {{% /tab %}}
 {{< /tabs >}}
 
+## Get self subject attributes
```
```suggestion
## API access to authentication information for a client {#self-subject-review}
```

nabokihms replied:
Fixed this one

sftim commented on content/en/docs/reference/access-authn-authz/authentication.md:
```diff
@@ -1239,6 +1239,49 @@ The following `ExecCredential` manifest describes a cluster information sample.
+
+With the `Self Subject Attributes Review` API, you can be aware of user attributes represented by your current environment (kubeconfig file or in-cluster config).
+
+For convenience, the `kubectl auth who-am-i` command is present. Executing this command will produce the following output (yet different user attributes will be shown):
```
Does this command also work with control plane versions before v1.26? If not, what happens when you try?

sftim commented on content/en/docs/reference/access-authn-authz/authentication.md:
```diff
@@ -1239,6 +1239,49 @@ The following `ExecCredential` manifest describes a cluster information sample.
+{{< /tabs >}}
+
+{{< note >}}
+The `APISelfSubjectAttributesReview` feature flag must be enabled to make this feature work.
```
Remember to add `APISelfSubjectAttributesReview` into https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/

nabokihms replied:
Added :+1:

sftim commented on content/en/docs/reference/access-authn-authz/authentication.md:
```diff
@@ -1239,6 +1239,49 @@ The following `ExecCredential` manifest describes a cluster information sample.
+
+{{< feature-state for_k8s_version="v1.25" state="alpha" >}}
+
+With the `Self Subject Attributes Review` API, you can be aware of user attributes represented by your current environment (kubeconfig file or in-cluster config).
```
```suggestion
You can use the `SelfSubjectReview` API, to find out how your Kubernetes cluster maps your authentication
information to identify you as a client. This works whether you are authenticating as a user (typically representing
a real person) or as a ServiceAccount.
```

nabokihms replied:
I applied your suggestion.

sftim commented on content/en/docs/reference/access-authn-authz/authentication.md:
````diff
@@ -1239,6 +1239,49 @@ The following `ExecCredential` manifest describes a cluster information sample.
+
+{{% tab name="YAML" %}}
+```yaml
+username: janedoe@example.com
+uid: b79dbf30-0c6a-11ed-861d-0242ac120002
+groups:
+- developers
+- system:authenticated
````
If there are extra data present, do these show up? If so, maybe we need a simple example and a more complicated example.

nabokihms replied:
I added a simple and a complicated example.

nabokihms commented:
@sftim Hello, I fixed the text according to you comments

- [ ] I would update https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/ to mention that you can use kubectl auth who-am-i, but only if your cluster has the feature enabled.

- [ ] We should document the format of a SelfSubjectReview and its fields (in https://k8s.io/docs/reference/access-authn-authz/authentication/ )

- [ ] We should document whether access mechanisms apply. Can everyone create a SelfSubjectReview, or do they need RBAC permissions on selfsubjectreviews before they can?

- [ ] How does this API interact with [impersonation](https://kubernetes.io/docs/reference/access-authn-authz/authentication/)?

- [ ] Can we mention when this API is especially useful (eg, in cluster that uses a custom authz webhook, or a cluster that uses an aggregated API that enforces additional authz).

Sea-n commented:
/remove-language zh

Sea-n commented:
/remove-area blog

nabokihms commented on content/en/docs/reference/access-authn-authz/authentication.md:
```diff
@@ -1241,6 +1241,142 @@ The following `ExecCredential` manifest describes a cluster information sample.
+        "editors"
+      ],
+      "extra": {
+        "provider_id": "token.company.dev"
```
This is a copypast from the KEP where you also requested to fix the same line. Sorry. I will fix this.

nabokihms commented on content/en/docs/reference/access-authn-authz/authentication.md:
```diff
@@ -1241,6 +1241,142 @@ The following `ExecCredential` manifest describes a cluster information sample.
+      "uid": "b6c7cfd4-f166-11ec-8ea0-0242ac120002",
+      "groups": [
+        "viewers",
+        "editors"
```
Yes, I will.

nabokihms commented on content/en/docs/reference/access-authn-authz/authentication.md:
```diff
@@ -1241,6 +1241,142 @@ The following `ExecCredential` manifest describes a cluster information sample.
+
+For convenience, the `kubectl alpha auth who-am-i` command is present. Executing this command will produce the following output (yet different user attributes will be shown):
+
+{{< tabs name="self_subject_attributes_review_Example_1" >}}
+{{% tab name="Simple output example" %}}
```
I split this into two sections: Simple / Complex and JSON / YAML. The first section is just a list, the second keeps the tabs formatting.

nabokihms commented on content/en/docs/reference/command-line-tools-reference/feature-gates.md:
```diff
@@ -58,6 +58,7 @@ different Kubernetes components.
 | `APIPriorityAndFairness` | `true` | Beta | 1.20 | |
 | `APIResponseCompression` | `false` | Alpha | 1.7 | 1.15 |
 | `APIResponseCompression` | `true` | Beta | 1.16 | |
+| `APISelfSubjectAttributesReview` | `false` | Alpha | 1.26 | |
```
My bad, fixed.

nabokihms commented on content/en/docs/reference/access-authn-authz/authentication.md:
```diff
@@ -1241,6 +1241,142 @@ The following `ExecCredential` manifest describes a cluster information sample.
+## API access to authentication information for a client {#self-subject-review}
+
+{{< feature-state for_k8s_version="v1.26" state="alpha" >}}
+
```
Added your variant.

nabokihms commented:
Fixed the doc according to all suggestions, thanks! Also added a not to the https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/ .

sftim commented:
/hold cancel

sftim commented:
New motivation for a hold: we should wait until the k/k PR actually merges.

@nabokihms what PR is that? (please link to this in the description for _this_ PR).

/hold

sftim commented on content/en/docs/tasks/access-application-cluster/configure-access-multiple-clusters.md:
````diff
@@ -389,6 +389,16 @@ export KUBECONFIG="$KUBECONFIG_SAVED"
 $Env:KUBECONFIG=$ENV:KUBECONFIG_SAVED
 ```
 
+## Check th subject represented by the kubeconfig
````
```suggestion
## Check the subject represented by the kubeconfig
```

sftim commented on content/en/docs/tasks/access-application-cluster/configure-access-multiple-clusters.md:
```diff
@@ -389,6 +389,16 @@ export KUBECONFIG="$KUBECONFIG_SAVED"
+
+There is a notable command to check subject attributes represented by contexts of your kubeconfig file - `kubectl alpha auth whoami`.
+
+[Here](/docs/reference/access-authn-authz/authentication/#self-subject-review) you can read about the feature in detail.
```
```suggestion
Read [API access to authentication information for a client](/docs/reference/access-authn-authz/authentication/#self-subject-review)
to learn about this in more detail.
```

sftim reviewed (commented):
A couple of suggestions.

nabokihms commented:
@sftim I updated the description (previously, there was a link to KEP with the same list of PRs).

The feature is merged into the main branch.

sftim commented:
/hold cancel

I recommend [squashing the commits](https://www.k8s.dev/docs/guide/github-workflow/#squash-commits) down to 1.

nabokihms commented:
@liggitt @sftim, is there anything left? Are we waiting for the review from one of the sig members?

liggitt commented:
technical content lgtm

sftim commented on content/en/docs/reference/access-authn-authz/authentication.md:
```diff
@@ -1241,6 +1241,147 @@ The following `ExecCredential` manifest describes a cluster information sample.
+{{< /tabs >}}
+
+This feature is extremely useful when a complicated authentication flow is used in a Kubernetes cluster, 
+for example, if you use [webhook token authentication](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#webhook-token-authentication) or [authenticating proxy](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#authenticating-proxy).
```
nits:
```suggestion
for example, if you use [webhook token authentication](/docs/reference/access-authn-authz/authentication/#webhook-token-authentication) or [authenticating proxy](/docs/reference/access-authn-authz/authentication/#authenticating-proxy).
```

sftim commented on content/en/docs/reference/access-authn-authz/authentication.md:
```diff
@@ -1241,6 +1241,147 @@ The following `ExecCredential` manifest describes a cluster information sample.
+
+{{< note >}}
+The Kubernetes API server fills the `userInfo` after all authentication mechanisms are applied,
+including [impersonation](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#user-impersonation).
```
nit:
```suggestion
including [impersonation](/docs/reference/access-authn-authz/authentication/#user-impersonation).
```

sftim commented on content/en/docs/reference/access-authn-authz/authentication.md:
```diff
@@ -1241,6 +1241,147 @@ The following `ExecCredential` manifest describes a cluster information sample.
+* the `APISelfSubjectReview`
+  [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
+  is enabled for your cluster
+* the API server for your cluster has `authentication.k8s.io/v1alpha1`
```
```suggestion
* the API server for your cluster has the `authentication.k8s.io/v1alpha1`
```

sftim commented on content/en/docs/reference/command-line-tools-reference/feature-gates.md:
```diff
@@ -633,6 +634,7 @@ Each feature gate is designed for enabling/disabling a specific feature:
 - `APIServerIdentity`: Assign each API server an ID in a cluster.
 - `APIServerTracing`: Add support for distributed tracing in the API server.
   See [Traces for Kubernetes System Components](/docs/concepts/cluster-administration/traces-for-kubernetes-system-components) for more details.
+- `APISelfSubjectAttributesReview`: Enable the authentication API to see the requesting subject's authentication information. Required to use the `kubectl alpha auth whoami` command.
```
```suggestion
- `APISelfSubjectAttributesReview`: Activate the `SelfSubjectReview` API which allows users
  to see the requesting subject's authentication information.
  See [API access to authentication information for a client](/docs/reference/access-authn-authz/authentication/#self-subject-review)
  for more details.
```

sftim commented on content/en/docs/tasks/access-application-cluster/configure-access-multiple-clusters.md:
```diff
@@ -389,6 +389,17 @@ export KUBECONFIG="$KUBECONFIG_SAVED"
+It is not always obvious what attributes (username, groups) you will get after authenticating to the cluster. 
+It can be even more challenging if you are managing more than one cluster at the same time.
+
+There is a notable command to check subject attributes represented by contexts of your kubeconfig file - `kubectl alpha auth whoami`.
```
```suggestion
There is a `kubectl` alpha subcommand command to check subject attributes, such as username,
for your selected Kubernetes client context: `kubectl alpha auth whoami`.
```

sftim reviewed (commented):
/lgtm

with a few nits that you can fix if you wish to @nabokihms

nabokihms commented:
@sftim Thank you for your suggestions. I applied them all.

sftim commented:
/lgtm

krol3 reviewed (approved):
LGTM cc @mickeyboxell

sftim commented on content/en/docs/reference/access-authn-authz/authentication.md:
```diff
@@ -1219,6 +1219,147 @@ The following `ExecCredential` manifest describes a cluster information sample.
+        "system:authenticated"
+      ],
+      "extra": {
+        "provider_id": ["token.company.dev"]
```
(nit; not a release blocker for alpha)
```suggestion
        "provider_id": ["token.company.example"]
```

`example` is a documentation top level domain; dev is a [real domain](https://get.dev/).

sftim commented on content/en/docs/reference/command-line-tools-reference/feature-gates.md:
```diff
@@ -377,6 +378,7 @@ Each feature gate is designed for enabling/disabling a specific feature:
 - `APIServerIdentity`: Assign each API server an ID in a cluster.
 - `APIServerTracing`: Add support for distributed tracing in the API server.
   See [Traces for Kubernetes System Components](/docs/concepts/cluster-administration/system-traces) for more details.
+- `APISelfSubjectAttributesReview`: Activate the `SelfSubjectReview` API which allows users
```
Should there be a line after this one?

sftim reviewed (commented):
/lgtm cancel

One thing to fix: the feature gate description looks truncated.

natalisucks commented:
@nabokihms Hello from the SIG Auth triage meeting! There's a nit above from Tim that would be great to get fixed before we merge; I'm happy to be the one to re-add an LGTM when the above is looked over 👍 Thanks so much!

nabokihms commented:
@sftim @natalisucks, sorry for the delay. I applied all suggestions.

krol3 commented:
Hi @nabokihms, This PR needs a doc review by **Mon Nov 28th** to get this into the release. Please reach out to required SIGs to get their review. Thank you!
cc: @sftim

natalisucks commented:
/lgtm
